third_party/spdk_defs/nvme_defs.h

```diff
@@ -0,0 +1,232 @@
+  // Reserved 0x17
+  kNamespaceAlreadyAttached = 0x18,
+  kNamespaceIsPrivate = 0x19,
+  kNamespaceNotAttached = 0x1a,
+  kThinprovisioningNotSupported = 0x1b,
+  kControllerListInvalid = 0x1c,
+  kDeviceSelfTestInProgress = 0x1,
+  kBootPartitionWriteProhibited = 0x1e,
+  kInvalidCtrlrId = 0x1f,
+  kInvalidSecondaryCtrlrState = 0x20,
+  kInvalidNumCtrlrResources = 0x21,
+  kInvalidResourceId = 0x22,
+
+  // NVM command set
+  kConflictingAttributes = 0x80,
+  kInvalidProtectionInfo = 0x81,
+  kAttemptedWriteToRoRange = 0x82,
+};
+
+// NVMe Base Specification Figure 130 and Figure 131
+// https://nvmexpress.org/wp-content/uploads/NVM-Express-1_4-2019.06.10-Ratified.pdf
+enum class MediaErrorStatusCode : uint8_t {
+  kWriteFaults = 0x80,
+  kUnrecoveredReadError = 0x81,
+  kGuardCheckError = 0x82,
+  kApplicationTagCheckError = 0x83,
+  kReferenceTagCheckError = 0x84,
+  kCompareFailure = 0x85,
+  kAccessDenied = 0x86,
+  kDeallocatedOrUnwrittenBlock = 0x87,
+};
+
+// NVMe Base Specification Figure 132
+// https://nvmexpress.org/wp-content/uploads/NVM-Express-1_4-2019.06.10-Ratified.pdf
+enum class PathStatusCode : uint8_t {
+  kInternalPathError = 0x00,
+  kControllerPathError = 0x60,
+  kHostPathError = 0x70,
+  kAbortedByHost = 0x71,
+};
+
+// NVMe Base Specification Figure 139 and Figure 140
+// https://nvmexpress.org/wp-content/uploads/NVM-Express-1_4-2019.06.10-Ratified.pdf
+enum class AdminOpcode : uint8_t {
+  kDeleteIoSq = 0x00,
+  kCreateIoSq = 0x01,
+  kGetLogPage = 0x02,
+  // Reserved 0x03
+  kDeleteIoCq = 0x04,
+  kCreateIoCq = 0x05,
+  kIdentify = 0x06,
+  // Reserved 0x07
+  kAbort = 0x08,
+  kSetFeatures = 0x09,
+  kGetFeatures = 0x0a,
+  // Reserved 0x0b
+  kAsyncEventRequest = 0x0c,
+  kNsManagement = 0x0d,
+  // Reserved 0x0e-0x0f
+  kFirmwareCommit = 0x10,
+  kFirmwareImageDownload = 0x11,
+  kDeviceSelfTest = 0x14,
+  kNsAttachment = 0x15,
+  kKeepAlive = 0x18,
+  kDirectiveSend = 0x19,
+  kDirectiveReceive = 0x1a,
+  kVirtualizationManagement = 0x1c,
+  kNvmeMiSend = 0x1d,
+  kNvmeMiReceive = 0x1e,
+  kDoorbellBufferConfig = 0x7c,
+  kFormatNvm = 0x80,
+  kSecuritySend = 0x81,
+  kSecurityReceive = 0x82,
+  kSanitize = 0x84,
+  kGetLbaStatus = 0x86,
+};
+
+// NVMe Base Specification Figure 346
+// https://nvmexpress.org/wp-content/uploads/NVM-Express-1_4-2019.06.10-Ratified.pdf
+enum class NvmOpcode : uint8_t {
+  kFlush = 0x00,
+  kWrite = 0x01,
+  kRead = 0x02,
+  // Reserved 0x3
+  kWriteUncorrectable = 0x04,
+  kCompare = 0x05,
+  // Reserved 0x06-0x07
+  kWriteZeroes = 0x08,
+  kDatasetManagement = 0x09,
+  kReservationRegister = 0x0d,
+  kReservationReport = 0x0e,
+  kReservationAcquire = 0x11,
+  kReservationRelease = 0x15,
+  kPdkNvmeSctGeneric = 0x0,
+  kSctCommandSpecific = 0x1,
+  kSctMediaError = 0x2,
+  kSctPath = 0x3,
+  // Reserved 0x4-0x6
+  kSctVendorSpecific = 0x7,
+};
+
+// NVMe Base Specification Figure 15
+// https://nvmexpress.org/wp-content/uploads/NVM-Express-1_0e.pdf
```

Comment:
Is the reason that you used Revision 1.0e because that's what spdk uses?

Reply:
Good question! I actually didn't consider what version I was linking here. I've noticed that all other links go to the 1_4 2019 spec of NVMe, but there's an updated 2020 spec 1_4a. I couldn't find a direct reference to what SPDK uses, but it may be the case that they update the library to always support features in the newest version. Is there a reason we chose the 2019 spec? https://nvmexpress.org/wp-content/uploads/NVM-Express-1_4a-2020.03.09-Ratified.pdf

Reply:
I believe the 2019 spec was listed in our project's google doc, so it might be worth asking the FTEs for this. But since the 2020 is just 1_4a, I don't imagine it'll be too different from the 2019 1_4 one

Reply:
I've been using the 2019 version to double check the SPDK implementation and then just leaving a comment where some fields were not included in the SPDK code

Reply:
Alright, I'll switch to using 1_4 for all my docs